### ISSUE TITLE
watchdog: Still fed when rootfs is non-responsive

### ISSUE THREAD

**Issue by Malvineous** [systemd/systemd#21083]
The systemd watchdog code does not check the filesystem when it feeds the watchdog.  This means that if there is a catastrophic failure that stops the root filesystem from responding, the whole machine becomes unusable, but systemd continues to feed the watchdog so the broken system never recovers.  In this situation it is not possible to log in to the machine (locally or remotely) and if there is an active session, no commands can be run as anything that tries to access the root filesystem simply freezes.

In my case I am running some Raspberry Pi devices with their root filesystems on NFS, and when I try to reboot them, the network goes away in the middle of the shutdown process.  This then causes the whole device to freeze waiting for the rootfs to return, which never happens.  The watchdog timer is active and should take over to reboot the machine as it does in other situations (such as a kernel panic) however in the dead-rootfs scenario systemd continues to feed the watchdog so the broken system never recovers and requires a power cycle to return to service, defeating the purpose of the watchdog.

I believe that systemd should read a non-cached file from the filesystem each time it tries to feed the watchdog, so a frozen rootfs would cause this process to also freeze, thus preventing the watchdog from being fed and causing a reset.

Note that the file being read cannot be cached from a previous read, as any files still cached do not result in a freeze.  And it must be on the root filesystem, as accesses to other filesystems like tmpfs still succeed.  Ideally reading the systemctl binary or other systemd-related files would be the best choice, provided any disk cache can be discarded before the read operation.

**systemd version the issue has been seen with**
 > 249.5

**Used distribution**
 > Arch Linux ARM

**Linux kernel version used** (`uname -a`)
 > `Linux 5.10.74-1-raspberrypi-ARCH #1 SMP Mon Oct 18 17:51:34 UTC 2021 armv7l GNU/Linux`

**CPU architecture issue was seen on**
 > aarch64

**Expected behaviour you didn't see**
 > System should have been rebooted by the watchdog when it stopped responding

**Unexpected behaviour you saw**
 > systemd kept feeding the watchdog even though the system was completely broken

**Steps to reproduce the problem**
1. Boot a machine with the watchdog timer active
2. Cause the root filesystem to fail (e.g. root on NFS and unplug network)
3. Observe the system is unusable (can't log in or run any programs) but despite this it never reboots

**Comment by yangm97** [comment]
Another scenario where this would be rather useful is when hardware i/o errors lock up the filesystem. I can almost reliably reproduce said "i/o issue" by using a knockoff micro sdcard as the rootfs. Issuing a trim command against one of these is almost certain to kill it in a way that, afaik, only a power cycle will take it out of its misery.

Despite this reproduction being for a very specific corner case, "more legitimate" environments do face similar issues, granted not as often and/or not as easy to reproduce. 

> Also note that in order to implement properly what you are looking for you probably need a statically linked binary that is locked into memory and not backed by files on disk anymore. i.e. the main daemon binary needs to be moved to a 'ramfs' or so, so that it is never backed by storage, and then executed from there, and then do its thing: once it detects hangs on the file systems it cares about it should first issue SIGRTMIN+15, and then maybe fall back to issuing reboot() itself after a timeout.

Say you have a service with `ExecStart=/bin/touch /.foo` and `FailureAction=reboot-immediate` being triggered constantly by a timer. Shouldn't it be able to "fail successfully" and trigger the reboot? 
I'm not well versed with the internals of systemd but, if the solution above is theoretically supposed to work, it would seen to me that systemd has all (or most) of the machinery needed to provide this functionality natively.

**Comment by trentbuck** [comment]
I use /dev/watchdog to detect when the rootfs (NFS squashfs + tmpfs cow) goes away or becomes corrupt (e.g. filesystem.squashfs is updated on the NFS server).
When this happens, often /sbin/reboot will not work because either /sbin/reboot isn't already cached in RAM, or (somehow) the system dbus is dead.  In such cases, ctrl+alt+del also does not work, because systemd can't find enough of itself to do a clean reboot.  In such cases, the system hangs forever - powered on, but unusable and unrebootable.

This is why /dev/watchdog is used, because it works even when the system is too damaged to reboot cleanly.

As a proxy for "is NFS dead?" I ping an httpd on the NFS server.
Very roughly, I do this:

    modprobe iTCO_wdt nowayout=1 heartbeat=60 &
    modprobe softdog  nowayout=1   timeout=60 &

    exec 99>/dev/watchdog
    echo pet >&99
    while curl -sSfL https://nfs
    do   echo pet >&99
         sleep 10
    done
    # try to do a clean reset
    reboot &
    # try to do a hard reset
    echo o >/proc/sysrq-trigger
    # wait and hope the watchdog will save us
    sleep infinity

This is sufficient for my needs even when using the in-kernel watchdog "softdog".

Re "It could be a simple service that every now and then checks if disk IO still works, and if not reboots.",
I do not see how this could work, because nfs (-o hard) blocks in D state (i.e. kernel-side).
The process cannot detect react to a dead NFS server because it cannot run until NFS comes back.

**Comment by trentbuck** [comment]
PS (getting off-topic): re "I believe that systemd should read a non-cached file from the filesystem" -- AFAIK the in-kernel linux NFS client caches pretty aggressively and doesn't provide any way to explicitly flush the cache of a single file.

Currently I do "umount -a -t nfs && mount -a -t nfs" to avoid around cache-related database corruption in sqlite journal_mode=WAL.  That MIGHT just be close-to-open caching, though -- I have not diagnosed it properly.
You could also do /proc/sys/vm/drop_caches which is system-wide -- worse even than just filesystem-wide.
This random quack result suggests an unnecessary opendir()+closedir() to flush CTO cache for a directory: https://stackoverflow.com/questions/8311710/nfs-cache-cleaning-command

**Comment by EternityForest** [comment]
On the Raspberry Pi it appears that a bad power supply can cause some kind of unrecoverable state of IO errors.  There are so many unusual combinations of failures that can happen, I don't think it makes sense to unnecessarily depend on anything but hardware.  

Any process you make that's supposed to catch whatever failure could itself fail, while the watchdog stuff could keep running. 


I'd like to see some kind of watchdog callback mechanism that lets you stop feeding it if a user-defined script, with a reasonable default, fails.

**Comment by Malvineous** [comment]
I think that's probably the most flexible solution.  Your health-check script could then check whatever is important to you (disk, network, external USB device, etc.) and respond OK if all is good and systemd can then feed the watchdog.  If your script fails to report in for any reason, then eventually the watchdog will expire.  Potentially you could have a number of `watchdog.d` type scripts, each one of which must report OK (or "not applicable") for the watchdog to be fed.

Only trick then is what to do during the shutdown/restart phase, e.g. if shutdown takes longer than the maximum watchdog expiry as it could on the Pi.

**Comment by EternityForest** [comment]
Yeah, the ability to have multiple scripts seems like a good idea.

I suppose you could have a new unit type for checkers that could specify when it's active, in terms of "Check this script as long as this other unit is intended to be running".

You could then do other nice to have features like allowing N failures before actually taking action.

You could only poll the units that actually had a watchdog flag enabled, so it could also be a general purpose alert system that other services could poll, to take action on less severe alerts,  or alerts that haven't been in the failed state long enough to reset.

**Comment by tomasz-torcz-airspace-intelligence** [comment]
No need for new unit type. We already have [FailureAction=](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#FailureAction=) which can reboot system on unit failure.

**Comment by EternityForest** [comment]
FailureAction would still require that the code to reboot actually be functional, right?

What about a new RequiredByWatchdog for timers?  That way the hardware can reset even if things are so corrupt that "sudo reboot now" would just give an IOError.

**Comment by Malvineous** [comment]
Yeah `FailureAction=` only works if the unit exits and cannot be restarted.  If the program freezes (due to a programming logic error, or a blocked syscall) then the process will never exit, and so `FailureAction=` will never trigger.  But the watchdog will keep being fed, and in this scenario, we want the watchdog to stop being fed when the process freezes.

**Comment by tomasz-torcz-airspace-intelligence** [comment]
If the program freezes, it will stop pinging unit watchdog (sd_notify) and systemd will trigger failure action.

**Comment by EternityForest** [comment]
I thought the unit watchdog was separate from the hardware watchdog?

In a corrupt state systemd might not be able to trigger anything, the
failure actions can involve a lot of complexity that might not be available
if the filesystem is gone or large parts of the kernel have been blasted by
a cosmic ray bit flip or similar.

You can't do a normal failure action in a state where you can't execute any
binaries or run any scripts at all.


This could all happen while the specific part of systemd that feeds the
watchdog keeps on going like everything is fine, because that part of the
code might just so happen to not be affected, and it currently doesn't
check for such things.

**Comment by pelzvieh** [comment]
I'd like to stress the aspect, that systemd seems to be continously feeding the watchdog (as in system.conf -> WatchdogDevice=...) even after it _did_ detect a catastrophic status.

What I'm seeing seems to hint, that systemd even feeds watchdog after executing its crash handler. I'm not absolutely sure, as I find it hard to debug – but I'm out of ideas why the hw watchdog wouldn't reboot the device when the last diagnostic send is like
`systemd[1]:  Caught <ABRT>, dumped core as pid 1234`
Instead the device is remaining on, the kernel still performing heartbeat light effect on a configured LED.

I'm trying now with system.conf option `CrashAction=reboot`, but cannot yet report if that's successful in that case.

Granted that it would be preferrable to eliminate the cause of the crash, but the point _here_ is that the watchdog mechanism is supposed to handle what hasn't been successfully handled, right?

I'm unsure what would be a less surprising solution, thinking along the lines:

If a watchdog is configured (= indicating the intent that the device should reboot out of a disfunctional state) `CrashAction=reboot` should be the default?

Moreovere, aren't some standard daemons, e.g. systemd-udevd, good candidates for a default `FailureAction=reboot`?
